torch/csrc/jit/codegen/cuda/tensor_view.cpp

```diff
@@ -423,6 +432,178 @@ TensorView* TensorView::computeAt(
+      getComputeAtPosition());
+
+  // If it's already set to be computed with the consumer and the
+  // position is higher, nothing to change
+  if (getComputeWithPosition() >= pos) {
+    return;
+  }
+
+  // Update the siblings together
+  auto siblings = ir_utils::filterByType<TensorView>(definition()->outputs());
+
+  for (auto sibling : siblings) {
+    sibling->clearComputeWith();
+  }
+
+  // If the given position is the same as the computeAt position, this
+  // is a no-op
+  if (pos == getComputeAtPosition()) {
+    return;
+  }
+
+  for (auto sibling : siblings) {
+    sibling->compute_with_pos_ = (unsigned int)pos;
+  }
+}
+
+bool TensorView::isComputedWith(const TensorView* consumer) const {
+  if (!hasComputeWith()) {
+    return false;
+  }
+
+  // Quering is an error if the compute-with consumer is still unresolved
+  TORCH_INTERNAL_ASSERT(
+      hasResolvedComputeWith(), "Not resolved yet: ", toString());
+
+  return std::find(
+             getComputeWithConsumers().begin(),
+             getComputeWithConsumers().end(),
+             consumer) != getComputeWithConsumers().end();
+}
+
+const std::vector<TensorView*>& TensorView::getComputeWithConsumers() const {
+  TORCH_INTERNAL_ASSERT(
+      !hasComputeWith() || hasResolvedComputeWith(),
+      "computeWith not yet resolved: ",
+      toString());
+  return compute_with_consumers_;
+}
+
+unsigned int TensorView::getMaybeComputeAtPosition(
+    const TensorView* consumer) const {
+  if (hasResolvedComputeWith() && isComputedWith(consumer)) {
+    return getComputeWithPosition();
+  } else {
+    return getComputeAtPosition();
+  }
+}
+
+bool TensorView::resolveComputeWith(const std::vector<Expr*>& sorted_exprs) {
+  TORCH_INTERNAL_ASSERT(
+      container()->isA<kir::Kernel>(), "Function invalid for fusion.");
+
+  auto siblings = ir_utils::filterByType<TensorView>(definition()->outputs());
+
+  for (auto sibling : siblings) {
+    TORCH_INTERNAL_ASSERT(
+        sibling->hasComputeWith(),
+        "Invlaid attempt to resolve computeWith: ",
+        sibling->toString());
+  }
+
+  // It may have been already resolved through its siblings
+  if (hasResolvedComputeWith()) {
+    return false;
+  }
+
+  std::unordered_set<Expr*> use_set;
+  for (auto sibling : siblings) {
+    use_set.insert(sibling->uses().begin(), sibling->uses().end());
+  }
+
+  for (auto expr : sorted_exprs) {
+    if (!use_set.count(expr)) {
+      continue;
+    }
+
+    // First use found. Set it as the computeWith target tensor
+    std::cerr << "Resolve the computeWith target as: " << expr->toString();
+    std::vector<TensorView*> use_out_tvs{
+        ir_utils::filterByType<TensorView>(expr->outputs()).begin(),
+        ir_utils::filterByType<TensorView>(expr->outputs()).end()};
+
+    for (auto sibling : siblings) {
+      sibling->compute_with_consumers_ = use_out_tvs;
+    }
+
+    for (auto consumer_tv : compute_with_consumers_) {
+      consumer_tv->updateMaxProducerPosition();
```

Comment:
Updating the max producer position of all consumers doesn't seem quite right to me. Except for one consumer where the producer is inlined, the other consumers still see the producer produced as if no computeWith is done. 

I don't remember what parts of the system would be affected by the change of the max producer position, but the expr sorting is one of them. Thinking about the normalization pattern, and I don't think it would be able to resolve the dependencies of the consumer and producer positions if we update the max producer position of all the consumers.

Reply:
I would assume it would have to be `maxProducerAtPosition` and `maxProducerWithPosition`.

Reply:
Yes, I'm thinking about adding something like `maybe_max_produce_position_`, which should work as a guard to prevent further transformations, it shouldn't be used for the expr sorting.